drivers: display: Add support of Seeed 2.8" TFT Touch Shield v2.0

leewkb4567 opened the pull request:
Add to ILI9340 display driver the support of Seeed 2.8" TFT Touch Shield v2.0. Support also RGB565 pixel format.
Add 2 nRF52 DK sample configurations to the ili9340 sample.

leewkb4567 commented:
@vanwinkeljan A new file 
'dts/bindings/display/ilitek,ili9341.yaml' is added. How is a code owner assigned?

vanwinkeljan commented:
> @vanwinkeljan A new file
'dts/bindings/display/ilitek,ili9341.yaml' is added. How is a code owner assigned?

Code owners are defined in [CODEOWNERS](https://github.com/zephyrproject-rtos/zephyr/blob/master/CODEOWNERS) feel free to add me and/or your self for ```/dts/bindings/display/```


Had no time yet to have a look at the code but are the ili9340 & ili9341 not very similar devices?
If this is the case could we not merge the 2 drivers?

leewkb4567 commented:
The PR has CODEOWNERS updated. Thanks.

The existing ili9340 driver supports RGB888 format. This ili9341 driver supports RGB565 format. Merging the 2 drivers will add further complexity.

vanwinkeljan reviewed (changes requested):
> The existing ili9340 driver supports RGB888 format. This ili9341 driver supports RGB565 format. Merging the 2 drivers will add further complexity.

Both controllers support RGB888 and RGB565 pixel format, further both controllers have exactly the same command set. So better add RGB565 support to existing ILI9340 driver and update documentation (Kconfig) that both controllers are supported.

leewkb4567 commented:
The controllers have minor differences in the command set. The command numbers C0h and C1h take different parameters. The display driver can handle the difference though.

The existing ILI9340 driver can indeed be modified to support 2 more ILI9341 LCD shields. The existing ILI9340 shield keeps working in RGB888 format. The ILI9341 shields work in RGB565 format. The ILI9340 driver LCD works in landscape orientation. The ILI9341 driver supports the LCD in portrait orientation and it will be handled as well.

vanwinkeljan commented:
> The existing ILI9340 driver can indeed be modified to support 2 more ILI9341 LCD shields. The existing ILI9340 shield keeps working in RGB888 format. The ILI9341 shields work in RGB565 format.

Would it not be better to add a KConfig in this case so that the user can pick its preferred option? In the end the RGB888 pixel gets reduced to 18-bit where the RGB565 gets extended through a LUT

> The ILI9340 driver LCD works in landscape orientation. The ILI9341 driver supports the LCD in portrait orientation and it will be handled as well.

Actually both controllers support landscape and portrait, the ILI9340 just picked landscape as default and the orientation can be changed through the API.

leewkb4567 commented:
> > The existing ILI9340 driver can indeed be modified to support 2 more ILI9341 LCD shields. The existing ILI9340 shield keeps working in RGB888 format. The ILI9341 shields work in RGB565 format.
> 
> Would it not be better to add a KConfig in this case so that the user can pick its preferred option? In the end the RGB888 pixel gets reduced to 18-bit where the RGB565 gets extended through a LUT

Not having an Adafruit 1480 shield to test the ILI9340 code, I will not modify the ILI9340 part.

Translating RGB565 to RGB888 by LUT is slow. The software has to translate every pixel. Not preferred.

I can try adding RGB888 support to the ILI9341 code. The register 3Ah format is the same.

The RGB565 format is preferred. It is more efficient. Each pixel takes 2 bytes instead of 3 bytes taken by RGB888 format. The SPI data transfer does not need padding too. The Arduino ILI9341 LCD vendors provide sample code in RGB565 format.

> > The ILI9340 driver LCD works in landscape orientation. The ILI9341 driver supports the LCD in portrait orientation and it will be handled as well.
> 
> Actually both controllers support landscape and portrait, the ILI9340 just picked landscape as default and the orientation can be changed through the API.

Somehow the two ILI9341 LCD shields take different register 36h values to support orientation. Again, not modifying the ILI9340 part because there is no hardware for testing.

vanwinkeljan commented:
> Not having an Adafruit 1480 shield to test the ILI9340 code, I will not modify the ILI9340 part.

I could test the code on the ILI9340

> Translating RGB565 to RGB888 by LUT is slow. The software has to translate every pixel. Not preferred.

What I mean is that the display controller it self uses a HW based LUT to translate RGB565 pixel (16-bit) to 18-bit pixel, so no need for a SW LUT.

> Somehow the two ILI9341 LCD shields take different register 36h values to support orientation. Again, not modifying the ILI9340 part because there is no hardware for testing.

I could test on the ILI9340 LCD shield, could be that the orientation code needs to be moved to display specific files in this case.

leewkb4567 commented:
The Adafruit LCD shields 1480 and 1947 are supported by the same Adafruit ILI9341 sample code. Therefore, the current ILI9340 driver should work on the 1947 shield. Nonetheless, it is not the case. The lvgl demo display is messed up.

For ILI9341, supporting RGB888 format cannot be done by simply writing 56h or 66h to register 3Ah. It applies to both Adafruit and Seeed shields.

On the contrary, the current ILI9340 driver can be easily modified to RGB565 format by writing 55h to register 3Ah. It works for both ILI9341 shields.

I will support only RGB565 format for ILI9341 shields. I do not have the resource to make RGB888 format.

Anyway, I will add ILI9341 shileds to the current ILI9340 driver.

leewkb4567 committed:
display: ili9340: Make reset gpio definition optional.

ILI9340 may not need a reset signal pin for certain LCD shields.

Signed-off-by: Bernard Lee <bernard.lee@nordicsemi.no>

vanwinkeljan commented:
> The Adafruit LCD shields 1480 and 1947 are supported by the same Adafruit ILI9341 sample code. Therefore, the current ILI9340 driver should work on the 1947 shield. Nonetheless, it is not the case. The lvgl demo display is messed up.

Hmm strange, would also assume that it no issue

> I will support only RGB565 format for ILI9341 shields. I do not have the resource to make RGB888 format.

Ok no problem to disable RGB888 for ILI9341, I will see if I can get my hands one of these shields

jfischer-no reviewed (changes requested):
I see no reason to add a new driver instead of extending the `drivers/display/display_ili9340.c` I tried this [ILI9341 based display](https://www.amazon.de/gp/product/B01E0MGZU2) a few months ago with  `samples/display/ili9340`, it worked out of the box.

leewkb4567 commented:
> I see no reason to add a new driver instead of extending the `drivers/display/display_ili9340.c` I tried this [ILI9341 based display](https://www.amazon.de/gp/product/B01E0MGZU2) a few months ago with `samples/display/ili9340`, it worked out of the box.

I will rework the patch to extend the ILI9340 driver. Not arguing for the current patch.

The ILI9340 / ILI9341 display controller is only part of the story. The LCD shield vendors may choose different LCD panels to suit their design requirements. Different LCD panels may support different signal formats. You guys have successes with two 2.2" shields. I have 2 failures with two 2.8" shields.

The Adafruit ILI9340 / ILI9341 shields come with a sample code that supports the 16-bit RGB565 format. It is the RGB format known working when these guys design a set of LCD shields. To mandate 18-bit RGB666 format, we may need to invest a lot of efforts to get every other similar modules working. What if we meet a LCD vendor that does not support the 18-bit format?

Does your Arduino shield come with a sample code? If yes, which RGB format is supported by the sample code? 18-bit or 16-bit? Can you give me a URL so that I can also take a look?

vanwinkeljan commented:
@leewkb4567 

> The Adafruit ILI9340 / ILI9341 shields come with a sample code that supports the 16-bit RGB565 format. It is the RGB format known working when these guys design a set of LCD shields. To mandate 18-bit RGB666 format, we may need to invest a lot of efforts to get every other similar modules working. What if we meet a LCD vendor that does not support the 18-bit format?

Note that the RGB formats defined in the display driver API are the format used on the SPI bus and not the format of the RGB interface between the display controller and the LCD. 

So in case of a display with 16-bit RGB interface to value of register 0x3A should be 0x56 for RGB888 and 0x55 for RGB565.

leewkb4567 committed:
samples: display: ili9340: Get ILI9340 device by DT label.

Get ILI9340 device by DT label instead of a fixed name.

Signed-off-by: Bernard Lee <bernard.lee@nordicsemi.no>

leewkb4567 committed:
samples: display: ili9340: ST NUCLEO-L476RG config file.

Put ST NUCLEO-L476RG specific parameters to own configuration file.

Signed-off-by: Bernard Lee <bernard.lee@nordicsemi.no>

leewkb4567 committed:
samples: display: ili9340: Add 2 Arduino shield sample configurations.

Add 2 Arduino shield sample configurations.
1. nRF52 DK (PCA10040) with Seeed 2.8" TFT Touch Shield v2.0.
2. nRF52840 DK (PCA10056) with Adafruit 2.8" TFT Touch Shield (1947).

Signed-off-by: Bernard Lee <bernard.lee@nordicsemi.no>

leewkb4567 commented:
It turns out that both ILI9341 LCD shields support 18-bit RGB. Register 0x3A is written a value of 0x66.

It is the display ili9340 sample working properly with the ILI9341 shields in RGB888 format.

The gui lvgl sample still does not work in RGB888 format. On the other hand, the RGB565 format works OK. It should not be a display driver issue.

The patch is updated moments ago. Please review again. Thanks.

vanwinkeljan commented on drivers/display/display_ili9340_seeed_tftv2.c:
```diff
@@ -0,0 +1,252 @@
+
+/**@brief Function to wait milliseconds.
+ */
+static void wait_ms(u16_t t)
```
Use k_sleep directly instead of wrapping it in an extra function.

leewkb4567 replied:
Will do.

vanwinkeljan commented on samples/gui/lvgl/boards/nrf52_pca10040.conf:
```diff
@@ -0,0 +1,14 @@
+
+CONFIG_ILI9340=y
+CONFIG_ILI9340_RGB565=y
+CONFIG_ILI9340_RGB565_SWAP=y
```
CONFIG_LVGL_COLOR_16_SWAP can be used to achieve the same functionality without the need of an extra buffer.

leewkb4567 replied:
When LittlevGL has RGB565 byte swap enabled, it becomes slower. On nRF52 devices, the LVGL full-screen processing time may be increased from 70ms to 100ms. On the other hand, the ILI9340 RGB565 byte swap just adds an overhead of 5ms.

Can we keep the ILI9340 RGB565 byte swap feature? Some users would like to boost the processing speed by giving a bit more RAM. It is also my preference.

vanwinkeljan replied:
Pretty surprised the overhead of lvgl is that muche for a byte swap. Still I would move the functionality out of the driver, why would we increase the complexity of a driver to solve a library issue?

Further on lvgl "issue" it could be an option to define a ```zephyr_vdb_write``` function that takes care of the swap in [lib/gui/lvgl/lvgl_color_16.c](https://github.com/zephyrproject-rtos/zephyr/blob/master/lib/gui/lvgl/lvgl_color_16.c) instead of using the lvgl swap implementation.

leewkb4567 replied:
Let's handle LittlevGL RGB565 byte swap in another PR.

When CONFIG_LVGL_COLOR_16_SWAP is enabled, LittlevGL suffers a compilation error. Need to disable CONFIG_LVGL_OBJ_CANVAS to build the library.

leewkb4567 replied:
The patch is updated. Your comments are addressed. Please review. Thanks.

leewkb4567 commented on drivers/display/display_ili9340_adafruit_1480.c:
```diff
@@ -27,8 +33,13 @@ void ili9340_lcd_init(struct ili9340_data *data)
 
+#ifdef CONFIG_ILI9340_RGB565
+	tx_data[0] = ILI9340_DATA_PIXEL_FORMAT_MCU_16_BIT |
+		     ILI9340_DATA_PIXEL_FORMAT_RGB_16_BIT;
```
The Adafruit sample code initializes the 1480 shield with 16-bit format. Let's take the easy way and follow suit.

In the 1480 shield sample code:
  https://github.com/adafruit/Adafruit_ILI9340/blob/master/Adafruit_ILI9340.cpp

  writecommand(ILI9340_PIXFMT);
  writedata(0x55);

In the later sample code for shields 1480, 1651, 1947:
  https://github.com/adafruit/Adafruit_ILI9341/blob/master/Adafruit_ILI9341.cpp

 ILI9341_PIXFMT  , 1, 0x55,

leewkb4567 commented on drivers/display/display_ili9340_adafruit_1480.c:
```diff
@@ -79,3 +90,54 @@ void ili9340_lcd_init(struct ili9340_data *data)
+		      const void *buf)
+{
+#ifdef CONFIG_ILI9340_RGB565
+	return ili9340_write_rgb565(data, desc, buf);
```
The different LCD shields have the identical write() function. I can use the same code.

For ILI9341 LCD, most users opt for either RGB565 or RGB888 format. There is little need for users to switch to and fro the RGB formats. How about waiving the pixel change feature and saving a bit of code space?

vanwinkeljan replied:
As we currently do not support changing the pixel format at run time in this driver I have no problem that de pixel byte size is defined at compile time.

leewkb4567 commented on drivers/display/display_ili9340.h:
```diff
@@ -61,4 +77,48 @@ void ili9340_transmit(struct ili9340_data *data, u8_t cmd, void *tx_data,
+ * @param data Device data structure
+ * @param capabilities Pointer to capabilities structure to populate
+ */
+void ili9340_lcd_get_capabilities(struct ili9340_data *data,
```
The two panels can use the same pixel formats. The write() and set_pixel_format() can be shared.

The Seeed panel LCD orientation code is different from the Adafruit one.

The Adafruit panel has value 0x28 written to register 0x36. In the same orientation, the Seed panel takes a value 0xB8.

The Seeed TFTv2 sample code initializes the panel in 240x320 portrait orientation. I tend to follow the vendor sample for a better viewing angle.

The current ILI9340 Adadruit driver puts the panel in 320x240 landsacpe orientation. A number of Zephyr samples are written in this configuration. Modifying the default orientation breaks the sample.

Anyway, the set_orientation() function is LCD specific.

vanwinkeljan replied:
On the orientation functionality this is not panel specific, the 0 coordinates are always the same for every panel. Problem here is the assembly of the panel in the final product (shield in this case) which could require calling the set orientation function at runtime. Alternatively a default orientation could be given per panel though KConfig.

leewkb4567 replied:
The Seeed LCD orientation will be aligned to Adafruit. The LCD specific orientation code are removed.

leewkb4567 commented on drivers/display/Kconfig.ili9340:
```diff
@@ -23,6 +23,36 @@ choice
+
 endchoice
 
+config ILI9340_RGB565_SWAP
```
Will do.

leewkb4567 committed:
display: ili9340: Adafruit LCD 1480 driver supports RGB565 pixel format.

Adafruit LCD 1480 driver supports RGB565 pixel format.

Signed-off-by: Bernard Lee <bernard.lee@nordicsemi.no>

vanwinkeljan commented:
please rebase

leewkb4567 committed:
display: ili9340: Add RGB565 pixel format support.

Add support of RGB565 pixel format to ILI9340 driver.

Signed-off-by: Bernard Lee <bernard.lee@nordicsemi.no>

leewkb4567 committed:
samples: display: ili9340: Add support of RGB565 pixel format.

ILI9340 display demo supports RGB565 as well as RGB888 pixel format.

Signed-off-by: Bernard Lee <bernard.lee@nordicsemi.no>

leewkb4567 commented:
> please rebase

The patch is rebased to solve the conflicts. The ili9340 sample is also adjusted.

vanwinkeljan commented:
@jfischer-phytec-iot could you have a second look at this PR?

jfischer-phytec-iot commented on drivers/display/display_ili9340_seeed_tftv2.c:
```diff
@@ -0,0 +1,169 @@
+
+	k_sleep(5);
+
+	cmd = 0xCFU;
```
@leewkb4567 @vanwinkeljan Are you happy with an other file full of magics? The command should be added to display_ili9340.h and readable defines used here. The same for data[.] magics, please add the register or register flags. This should be readable and understandable by humans.

leewkb4567 replied:
The Seeed LCD shield works when I apply their sample code to nRF52 DK. Actually I do not understand their command sequence either. The LCD shield vendors rarely offer any documentation to program their display.

Anyway, I will update this Seeed driver code to make it easier to read.

leewkb4567 replied:
The Seeed LCD initialization has been updated with ILI9341 command numbers.

Some command numbers are not found in ILI9341 datasheet V1.02. They are found in newer datasheet V1.11. A PDF copy can be located in following URL:

https://www.newhavendisplay.com/app_notes/ILI9341.pdf

Please review again. Thanks.

leewkb4567 committed:
display: ili9340: Add support of Seeed 2.8" TFT Touch Shield v2.0.

Add support of Seeed 2.8" TFT Touch Shield v2.0 to ILI9340 driver.
This driver supports color pixel formats of RGB888 and RGB565.

Signed-off-by: Bernard Lee <bernard.lee@nordicsemi.no>

pabigot reviewed (dismissed):
OK, this does work, but I will ask that the changes to the SPI pins from the board standard assignments in DTS be removed, unless there is a compelling reason why this example can't use the standard assignments.

Also see https://github.com/zephyrproject-rtos/zephyr/issues/16428#issuecomment-496325511 because it's not clear whether the remaining problem with that sample is with the lvgl layer or with this driver.

leewkb4567 commented on samples/display/ili9340/nrf52_pca10040.overlay:
```diff
@@ -0,0 +1,22 @@
+
+&spi1 {
+	status = "ok";
+	sck-pin = <25>;
```
In this sample, the Seeed LCD shield is plugged in the Arduino-compatible connector on the nRF52 DK (PCA10040). These GPIO pins are selected to match the hardware configuration.

leewkb4567 commented on samples/display/ili9340/nrf52840_pca10056.overlay:
```diff
@@ -0,0 +1,21 @@
+ */
+
+&spi1 {
+	status = "ok";
```
In this sample, the Adafruit LCD shield 1947 is plugged in the Arduino-compatible connector on the nRF52840 DK (PCA10056). These GPIO pins are selected to match the hardware configuration.

For the 'standard' pins, the user has to connect the two boards by wiring.

pabigot replied:
OK, so either the Nordic boards fail to specify a SPI driver that aligns to Arduino pins, which is not your problem, or the shields don't match what Nordic provided.  In either case, please document in the overlay that these specific pins were chosen to support Arduino shields, and maybe provide the readable names  (P1.15 instead of 47).

Also note that in this case RESET needs to be tied to Vdd since there isn't a gpios property for it.  Somebody could spend a lot of time trying to figure out what's wrong if that signal isn't connected on the Adafruit display.

Most of this documentation really belongs in the README, with just a reference to Arduino shield compatibility in the overlay.

leewkb4567 replied:
The README will be updated for the additional hardware configurations. The Zephyr documentation also needs to be updated. If possible, I will make it in another PR.

The Adafruit LCD shield 1947 works without a RESET pin driven by the ili9340 driver. Will give the GPIO details in README.

leewkb4567 commented:
> OK, this does work, but I will ask that the changes to the SPI pins from the board standard assignments in DTS be removed, unless there is a compelling reason why this example can't use the standard assignments.
> 
> Also see [#16428 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/16428#issuecomment-496325511) because it's not clear whether the remaining problem with that sample is with the lvgl layer or with this driver.

The comment on GPIO pins assignments are addressed separately. 

I could not get the my ILI9341 LCD shields working properly with the lvgl in 24-bit format either. I guess the lvgl sample is somehow broken at some point. I could get the lvgl working in 16-bit format though. You have to configure both ili9340 and lvgl to work in 16-bit format. The screen dimensions also need to match. Nevertheless, ili9340 lvgl 16-bit format has a performance issue that I would like to tackle later. It is why I use instead the ili9340 sample to show that the ili9340 driver is working.

pabigot commented on drivers/display/display_ili9340_seeed_tftv2.c:
```diff
@@ -0,0 +1,174 @@
+	u8_t cmd;
+	u8_t data[15];
+
+	/* Software reset */
```
Most display drivers I've seen use a technique like [this](https://github.com/adafruit/Adafruit_ILI9341/blob/master/Adafruit_ILI9341.cpp#L148) to specify the commands and arguments, which is shorter and more readable than a series of assignments into an array.  Would you consider moving to that form?

jfischer-phytec-iot replied:
> Most display drivers I've seen use a technique like [this](https://github.com/adafruit/Adafruit_ILI9341/blob/master/Adafruit_ILI9341.cpp#L148) to specify the commands and arguments, which is shorter and more readable than a series of assignments into an array. Would you consider moving to that form?

Exactly like [this](https://github.com/adafruit/Adafruit_ILI9341/blob/master/Adafruit_ILI9341.cpp#L148) should not be used. It is absolutely not clear what is initialized there.

pabigot replied:
> Exactly like [this](https://github.com/adafruit/Adafruit_ILI9341/blob/master/Adafruit_ILI9341.cpp#L148) should not be used. It is absolutely not clear what is initialized there.

Well, no, certainly the commands which have identified names like `ILI9340_CMD_VCOM_CTRL_1` should be expressed using those names.  But most of the data values are those provided by the display manufacturer as appropriate for the way they built the device.  I'm personally fine with `0x55` vs `0x66` for `ILI9340_CMD_PIXEL_FORMAT_SET` rather than constructing it from constants like `ILI9340_DATA_PIXEL_FORMAT_RGB_16_BIT`, because the two or three people who will need to know what the value means over the maintainable life of the driver can reference the datasheet, which they'll probably need anyway.

I don't think it's worth going back and converting all the magic numbers to constructed values with named bitfields, except maybe in a case where the corresponding register can be changed at runtime, or if the chip vendor provided a header with all the constants defined in it so the mapping doesn't become another source of error.

Part of my position on this is because I spent a lot of time yesterday comparing SPI bus captures from the Zephyr application with the Adafruit reference implementation, which was a lot easier than comparing with the Zephyr code because I could immediately tell where the differences were (0x55 vs 0x66 was one of them).

But it's a choice of where you draw the line.

leewkb4567 replied:
Sorry. I do not have a plan to migrate to a table-style sequence now. The Seeed sample code is not in table-style and I port it plainly to Zephyr.

pabigot reviewed (commented):
These are the things that I would consider blocking if I had responsibility for this area, but I don't so I'm only expressing an opinion.

* This PR adds support for Arduino shield form-factor displays, and for Nordic platforms, but does not update the README.  The README only mentions support for Adafruit 2.2" TFT on NUCLEO which needs to be wired directly.  The README should be updated to describe the new capabilities.
* The Nordic overlays that replace the board DTS pin assignments should have a comment indicating that the assignments were selected to be compatible with Arduino shields.
* I would start an investigation into why two Nordic boards appear to have DTS files that are not compatible with Arduino shield use of SPI.
* The device initialization code is needlessly complex.  It's possible the compiler converts all those assignments to references to a static const byte array, but it would be cleaner if the sequence implemented using the technique identified in the corresponding review comment.

Thank you for your work on this.  I have no interest in this display, but did want a working example of LVGL in Zephyr before trying to support other displays, and this went a long way towards that goal.

leewkb4567 commented:
The Nordic nRF51 and nRF52 series SoC can let SPI, TWI and UART blocks working on any of the GPIO pins. The only precaution is that these GPIO pins should not be near the radio antenna. I think the nRF52 DK DTS files has selected the SPI GPIO pins arbitrarily. The driver samples can modify the SPI GPIO pin assignments with the overlay files.

For lvgl, the maintainer will take a look at issue #16428.
